### ISSUE TITLE
Traefik.docker.network doesn't work with 2 bridge networks on docker-compose

### ISSUE THREAD

**Issue by RealKanashii** [containous/traefik#4312]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

Hi, I want to be able to create two different networks on traefik with docker-compose. The first one is frontend where I want to serve the client request and at the same time make another one to use as internal communication but excluding traefik. So some of my dockers are in both networks ( ex: grafana ) and some only in the frontend network. But no matter what I do, I always see on traefik's status the dockers using both network randomly as if `traefik.docker.network` was not working.

### What did you expect to see?

All the docker on the backends with the same label `traefik.docker.network:gestio_frontend` on the same network.

### What did you see instead?

Randomly assigned networks on backends:

<details>

![image](https://user-images.githubusercontent.com/9731708/50158542-b258d800-02d4-11e9-8221-6dc71e17e21e.png)

</details>

### Output of `traefik version`: (_What version of Traefik are you using?_)

```console
jlorenzo@docker01:~/gestio$ sudo docker run traefik version
Version:      v1.7.5
Codename:     maroilles
Go version:   go1.11.2
Built:        2018-12-03_11:01:00AM
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

**Docker-compose bin**
```console
jlorenzo@docker01:~/gestio$ sudo docker-compose version
docker-compose version 1.23.1, build b02f1306
docker-py version: 3.5.0
CPython version: 3.6.7
OpenSSL version: OpenSSL 1.1.0f  25 May 2017
```
**Docker**
```console
jlorenzo@docker01:~/gestio$ sudo docker --version
Docker version 18.09.0, build 4d60db4
```

**SO Ubuntu 18** 
```console
jlorenzo@docker01:~/gestio$ uname -a
Linux docker01 4.15.0-42-generic #45-Ubuntu SMP Thu Nov 15 19:32:57 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

**traefik.toml**
```toml
defaultEntryPoints = ["http"]

[entryPoints]
  [entryPoints.http]
  address = ":80"

[docker]
endpoint = "unix:///var/run/docker.sock"
domain = "docker.localhost"
exposedByDefault = false

```
**docker-compose.yml**
```yml
version: "3.7"
networks:
  backend:
    driver: bridge
  frontend:
    driver: bridge
volumes:
  portainer_data:
  grafana_data:
services:
  influxdb:
    labels:
      - "traefik.enable=false"     
      - "traefik.frontend.rule=Host:influxdb.test.local;"  
    image: influxdb 
    volumes:
      - /var/lib/docker/influxdb:/var/lib/influxdb
    networks:
      - backend
  cadvisor:
    labels:
      - "traefik.enable=false"
      - "traefik.frontend.rule=Host:cadvisor.test.local;" 
    image: google/cadvisor 
    volumes:
      - /:/rootfs:ro # 
      - /var/run:/var/run:ro 
      - /sys:/sys:ro 
      - /var/lib/docker/:/var/lib/docker:ro 
    networks:
      - backend
      - frontend
    command: -storage_driver=influxdb -storage_driver_db=cadvisor_db -storage_driver_user=xxxxxxx -storage_driver_password=xxxxxx -storage_driver_host=influxdb:8086
    depends_on:
      - influxdb
  prometheus:
    labels:
      - "traefik.enable=true"
      - "traefik.docker.network:gestio_frontend"
      - "traefik.port=80"
      - "traefik.frontend.rule=Host:prometheus.test.local"
    image: prom/prometheus
    networks:
      - backend
    volumes:
      - /home/user/gestio/prometheus.yml:/etc/prometheus.yml
  grafana:
    labels:
      - "traefik.enable=true"  
      - "traefik.docker.network:gestio_frontend"
      - "traefik.port=80"
      - "traefik.frontend.rule=Host:monitor.test.local;"
    image: grafana/grafana
    environment:
      - GF_SECURITY_ADMIN_PASSWORD=XXXXX 
    volumes:
      - grafana_data:/var/lib/grafana  
    networks:
      - backend
      - frontend
  portainer:
    labels:
      - "traefik.enable=true"     
      - "traefik.docker.network:gestio_frontend"
      - "traefik.port=80"
      - "traefik.frontend.rule=Host:orquestrador.test.local;" 
    image: portainer/portainer
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - portainer_data:/data
    networks:
      - frontend
  traefik:
    image: traefik  # Imatge oficial de traefik
    command: --api --docker --web --web.metrics.prometheus --web.metrics.prometheus.buckets="0.1,0.3,1.2,5.0" logLevel=DEBUG
    ports:
      - "80:80"  
      - "443:443" 
      - "8080:8080" 
    networks:
      - frontend
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock  
      - /home/user/signasuite/traefik.toml:/etc/traefik/traefik.toml  docker
```
Docker's Networks
```console
jlorenzo@docker01:~/gestio$ sudo docker network ls
NETWORK ID          NAME                DRIVER              SCOPE
d6521d3d44c7        bridge              bridge              local
43686b370425        gestio_backend      bridge              local
288f68d62f2b        gestio_frontend     bridge              local
25475bc387ea        host                host                local
60b8fd783dfd        none                null                local
```

And the docker are on both network with inspect and the tag seems configured:
```console
jlorenzo@docker01:~/gestio$ sudo docker inspect 05788279801f | grep -i network
            "NetworkMode": "gestio_backend",
                "traefik.docker.network:gestio_frontend": "",
        "NetworkSettings": {
            "Networks": {
                    "NetworkID": "43686b3704251e79f52164a78b177e76dc92efd332afdbf0ee41d29b0c3180a1",
                    "NetworkID": "288f68d62f2b9ff17b5a428717b49b694f32a3074d752390cb28c9dd4adc071a",
```

I tried to use "default" and if I set the traefik on both networks then the reverse proxy works (as traefik sees both network and can route the request to backend) but I want the second one out of traefik network.

**Comment by RealKanashii** [comment]
Tell me if you need more info or debugging. I will try an older version of traefik docker to validate if it's a version related bug or not.

**Comment by RealKanashii** [comment]
Edited main post to fix the ident on two services (portainer & traefik). Was a cut&paste error, not a real config error (the docker-compose builds the conf perfectly).

**Comment by Eagle-One1** [comment]
Hello @RealKanashii, did you manage to use traefik with 2 networks ? 
I'm "building" around docker. With traefik and portainer, I wish to expose some services only on my local network, with ports 443 and 80, and with an other traefik  expose some other services on port 8001 and 8443 that I NAT with my box so that I can access only the services i wanted to expose out of my LAN over the internet, and the others I don't want to are only accessible on my LAN

**Comment by marema31** [comment]
Hello, I was experimenting with the same stack, and I encountered the same problem.

Traefik: 1.7.6
Docker: 18.09.0, build 4d60db4
Docker-compose: 1.23.2, build 1110ad01
Hosted on: Centos 7.6.1810

When I added the second network to traefik block, it works as expected.

**Comment by Eagle-One1** [comment]
Hello, 

Are the 2 networks separated ? 
With 2 traeficks on 2 networks, as bridge, one traefik on each networks, with one app on only one of the two networks but visibles by the 2 traeficks, not what I was expecting

**Comment by marema31** [comment]
The traefik compose file

`
version: '3.7'
services:
  traefik:
    image: traefik:latest
    command: --api --api.dashboard=true -c /dev/null --docker --logLevel=WARN --accessLog --docker.endpoint=unix:///var/run/docker.sock --docker.watch --entryPoints='Name:http Address::80' --defaultentrypoints=http --constraints=tag==web
    ports:
      - 80:80
    volumes:
      - "/var/run/docker.sock:/var/run/docker.sock"
      - /dev/null:/traefik.toml
    networks:
      - web
networks:
  web:
   name: traefik_web
  monitoring:
    external:
      name: monitoring
`

And part of the monitoring compose:
`
version: '3.7'
services:
  grafana:
    image: grafana/grafana:5.3.2
    environment:
      - GF_SERVER_URL=http://mon.lan.carmier.fr/
   networks:
      - traefik_web
      - monitoring
    labels:
      - "traefik.tags=web"
      - "traefik.port=3000"
      - "traefik.backend=mon"
      - "trafik.docker.network=traefik_web"
      - "traefik.frontend.rule=Host:mon.example.com"

networks:
  monitoring:
    name: monitoring
  traefik_web:
   external:
     name: traefik_web
`
If I had the monitoring network to traefik service, it works, if not I have a gateway timeout.

**Comment by RealKanashii** [comment]
@marema31 but you have no problem. You are using 2 stacks. Stacks are isolated by default so if you need a stack to be visible by traefik stack you need to share the traefik network on the other stack ("monitor in your case").
My problems is using 2 networks on the same stack and with containers using both stack networks. Seems traefik detects dockers but assings the incorrect IP even specifying the main frontend network.

**Comment by RealKanashii** [comment]
@Eagle-One1. That's exactly the problem. I cannot define two network at the same time (frontend and backend) in the same compose (stack) and present those dockers to traefik cause traefik ignores the "traefik.network" property and uses ramdomly one of the docker IPs ( sometimes is from frontend network and sometimes from backend) producing bad gateway error due bad backend assignment.

**Comment by marema31** [comment]
@RealKanashi, I have the same problem, I don't use swarm but just docker-compose on developper desk, the fact I had two stack is no different since I declare that some networks are "external".
On this machine, I have other containers launched by docker-compose (for simplicity for the users) that define the traefik_web as external and traefik is able to redirect to them without problem.
In the example I provided there is two way to make it functionnal:

- add monitoring network to traefik block
- remove monitoring network to grafana

But I would prefer that traefik use the "traefik_web" network as defined by the trafik.docker.network label.

**Comment by RealKanashii** [comment]
Ok, I got it now. I will try the same adding a docker run defined with two networks and see if the problem is the same (not only on compose). One of the networks from traefik and another isolated and validate the docker has randomly assigned backend IPs too.

**Comment by RealKanashii** [comment]
Any news? I consider have a backend network inside the stack witouth being exposed a must. So one stack can simulate 2 vlans ( service and management).

**Comment by si458** [comment]
`- "trafik.docker.network=traefik_web"`
shouldnt this read?
` - "traefik.docker.network=traefik_web"`

**Comment by marema31** [comment]
@si458 : you're right for the typo, I may have introduced during my tests, I just done some tries with correcting it and it does not worked, I've restarted several time the two stacks and it didn't work each time

**Comment by jgillich** [comment]
Same issue here, just wasted an hour trying to understand what was going on. traefik.docker.network does not work as expected.

**Comment by viettuan1807** [comment]
Any update on this?

**Comment by mpl** [comment]
@marema31 I'm only a little bit familiar with the code of the v2 (and not the v1.7), so maybe this is not relevant to your problem, but while trying to understand that issue, I noticed this part of the code which might shed some light:

https://github.com/containous/traefik/blob/v2.0/pkg/provider/docker/config.go#L236-L246

i.e. it seems to me that, for a given container, if a network foo is not declared in its "networks" section, and if one tries to use that foo network in a network label ("traefik.docker.network=foo"), then that label is ignored. Does that correspond with the behaviour you're experiencing?

in other words, are you seeing a log message such as:
"Could not find network named '%s' for container '%s'! Maybe you're missing the project's prefix in the label? Defaulting to first available network." ?
(again, maybe this log message is only in the v2, I haven't checked the v1.7 code).

**Comment by marema31** [comment]
Hello @mpl,
I didn't have this kind of message, but I'm not sure about the relevance because if you look the compose files I've posted before, the networks are all declared in networks: sections.

**Comment by RealKanashii** [comment]
> Hello @RealKanashii, did you manage to use traefik with 2 networks ?
> I'm "building" around docker. With traefik and portainer, I wish to expose some services only on my local network, with ports 443 and 80, and with an other traefik expose some other services on port 8001 and 8443 that I NAT with my box so that I can access only the services i wanted to expose out of my LAN over the internet, and the others I don't want to are only accessible on my LAN

Better use the traefik IP filter. Another solution is present a port on the internal network on the container ( with "port" in compose) but use another on the traefik with the label "traefik.port=XXXXX" where X is the external port. On internal network use the main port on the container. 

So what you want is simplier if you use the same network but diferrent ports pointing to traefik which port to use for external. When you don't define a port in traefik network traefik look for all known ports and route traffic to that ports. If you specify then it will use only the defined.

**Comment by RealKanashii** [comment]
> @marema31 I'm only a little bit familiar with the code of the v2 (and not the v1.7), so maybe this is not relevant to your problem, but while trying to understand that issue, I noticed this part of the code which might shed some light:
> 
> https://github.com/containous/traefik/blob/v2.0/pkg/provider/docker/config.go#L236-L246
> 
> i.e. it seems to me that, for a given container, if a network foo is not declared in its "networks" section, and if one tries to use that foo network in a network label ("traefik.docker.network=foo"), then that label is ignored. Does that correspond to the behaviour you're experiencing?
> 
> in other words, are you seeing a log message such as:
> "Could not find network named '%s' for container '%s'! Maybe you're missing the project's prefix in the label? Defaulting to first available network." ?
> (again, maybe this log message is only in the v2, I haven't checked the v1.7 code).

Ok, now you pointed it I see a misconfig in Prometheus declaration:

```
prometheus:
    labels:
      - "traefik.enable=true"
      - "traefik.docker.network:gestio_frontend" <----!!!
      - "traefik.port=80"
      - "traefik.frontend.rule=Host:prometheus.test.local"
    image: prom/prometheus
    networks:
      - backend <----!!!
    volumes:
      - /home/user/gestio/prometheus.yml:/etc/prometheus.yml
```
I supose you're talking about this. You're right, that container was misconfigured but the other container which is well defined on both compose network declaration and containers have the same behavior.

**Comment by mpl** [comment]
@marema31 hmm maybe the match is on the network name though? I notice that in the grafana container definition you've been consistent on that matter, i.e. you have used the "traefik_web" name both in the networks section of the container, and in the label. On the other hand, in the traefik container definition, you have used "web" in the networks section of the container, so maybe you should use "traefik_web" there too?

**Comment by gintsgints** [comment]
Any news on this?

**Comment by hellerchr** [comment]
@RealKanashii looks like a typo to me.

"traefik.docker.network:gestio_frontend" 
-> 
"traefik.docker.network**=**gestio_frontend"

**Comment by traefiker** [comment]
Hi! I'm Træfiker :robot: the bot in charge of tidying up the issues.

I have to close this one because of its lack of activity :disappointed:

Feel free to re-open it or join our [Community Forum](https://community.containo.us/).
